### ISSUE TITLE
mangos v2 dependency broken with dep

### ISSUE THREAD

**Issue by edermi** [nanomsg/mangos#79]
I'm using mangos v2 in a project where dependencies are managed via [dep](https://golang.github.io/dep/). This worked fine last week but running `dep ensure` today throws the following error message:

```
The following issues were found in Gopkg.toml:

  ✗ the name for "nanomsg.org/go/mangos/v2" should be changed to "nanomsg.org/go/mangos"

ProjectRoot name validation failed
```
So, I edited `Gopkg.toml`:

```
[[constraint]]
  name = "nanomsg.org/go/mangos/v2"
  version = "2.0.0"
```

and removed the "/v2":

```
[[constraint]]
  name = "nanomsg.org/go/mangos"
  version = "2.0.0"
```

This unfortunately introduces new problems when I try to update the dependencies using `dep ensure -update`:

```
Solving failure: No versions of nanomsg.org/go/mangos met constraints:
	v2.0.1: Could not introduce nanomsg.org/go/mangos@v2.0.1 due to multiple problematic subpackages:
	Subpackage nanomsg.org/go/mangos/v2/protocol/rep is missing. (Package is required by (root).)	Subpackage nanomsg.org/go/mangos/v2/protocol/req is missing. (Package is required by (root).)	Subpackage nanomsg.org/go/mangos/v2/transport/tcp is missing. (Package is required by (root).)	Subpackage nanomsg.org/go/mangos/v2 is missing. (Package is required by (root).)
	v2.0.0: Could not introduce nanomsg.org/go/mangos@v2.0.0 due to multiple problematic subpackages:
	Subpackage nanomsg.org/go/mangos/v2 is missing. (Package is required by (root).)	Subpackage nanomsg.org/go/mangos/v2/protocol/rep is missing. (Package is required by (root).)	Subpackage nanomsg.org/go/mangos/v2/protocol/req is missing. (Package is required by (root).)	Subpackage nanomsg.org/go/mangos/v2/transport/tcp is missing. (Package is required by (root).)
	v1.4.0: Could not introduce nanomsg.org/go/mangos@v1.4.0, as it is not allowed by constraint ^2.0.0 from project <myproject>.
	v1.3.0: Could not introduce nanomsg.org/go/mangos@v1.3.0, as it is not allowed by constraint ^2.0.0 from project <myproject>.
	v1.2.0: Could not introduce nanomsg.org/go/mangos@v1.2.0, as it is not allowed by constraint ^2.0.0 from project <myproject>.
	v1.1.0: Could not introduce nanomsg.org/go/mangos@v1.1.0, as it is not allowed by constraint ^2.0.0 from project <myproject>.
	v1: Could not introduce nanomsg.org/go/mangos@v1, as it is not allowed by constraint ^2.0.0 from project <myproject>.
	v1.0.0: Could not introduce nanomsg.org/go/mangos@v1.0.0, as it is not allowed by constraint ^2.0.0 from project <myproject>.
	master: Could not introduce nanomsg.org/go/mangos@master, as it is not allowed by constraint ^2.0.0 from project <myproject>.
```

It seems that now there is a namespace/version conflict with nanomsg 1 that does not allow me to upgrade my packages.

I see that there are currently some things moving around regarding go modules and dependency management and I'm also aware that `dep` is sort of an "official playground" but I need some kind of dependency management beyond `go get` that does not require latest go and works with existing code, and dep is a good and widely used choice.

Unfortunately I can't tell you what exactly broke dep / when exactly that happened, but it worked last week and I (+ lots of others I guess) would love to use mangos-v2 together with dep again.

**Comment by gdamore** [comment]
So I think this is partly my fault, as I made changes to accommodate the new (go1.11) modules system.

One thing is you should use the latest version of go 1.9, 1.10, or 1.11.  (Specifically there were updates made to 1.9 and 1.10 to make this work.)  I’m not sure if this will be sufficient for you, as I don’t know if “dep” fixes this.  You might also check to see if the “dep” code has updates to work better… I suspect not.

I may wind up reverting those changes, in which case I’ll probably need to start a mangos 3 project to accommodate the changes in go 1.11 later… I hope this won’t be necessary.

More on “dep” vs “mod” here: https://github.com/golang/dep/issues/1959

The whole thing makes me want to cry…..

 - Garrett

**Comment by edermi** [comment]
I'm currently using Go 1.11, but I can't assume that version to be generally available on all machines - assuming 1.9 as lowest should be fine though. I'm also running latest dep (the release is from July), so I guess the problem did not go away over night. I see that you are getting into quite some trouble regarding this topic and I guess there will be more issues in the future. For now, It seems that I will have to give go modules a shot.

Nevertheless, thanks for your reply and I hope you find a good solution to solve this. You may close the issue if you wish.

**Comment by gdamore** [comment]
I think #81 covers this.  dep is the old way, and there were bugs there, as well as in go modules.  This is geting fixed in go itself, in the meantime I can't really offer much wisdom here.  Sorry.
